File: bot/keyboards.py
"""Keyboard builders for inline buttons."""

from __future__ import annotations

from typing import Iterable, List, Optional, Sequence, Set

from telegram import InlineKeyboardButton, InlineKeyboardMarkup

from config import settings
from . import messages
from .questions import Option, Question


def _build_inline_keyboard(rows: Sequence[Sequence[InlineKeyboardButton]]) -> InlineKeyboardMarkup:
    return InlineKeyboardMarkup(list(rows))


def skill_level_keyboard() -> InlineKeyboardMarkup:
    buttons: List[List[InlineKeyboardButton]] = []
    for callback_data, text in messages.SKILL_LEVEL_OPTIONS:
        if not text:
            continue
        buttons.append([InlineKeyboardButton(text=text, callback_data=callback_data)])
    return _build_inline_keyboard(buttons)


def start_keyboard() -> InlineKeyboardMarkup:
    return single_button_keyboard(messages.START_BUTTON)


def single_button_keyboard(button_data: tuple[str, str]) -> InlineKeyboardMarkup:
    callback_data, text = button_data
    return _build_inline_keyboard([[InlineKeyboardButton(text=text, callback_data=callback_data)]])


def question_options_keyboard(
    question: Question,
    selected_options: Set[str] | None = None,
) -> InlineKeyboardMarkup:
    selected_options = selected_options or set()
    rows: List[List[InlineKeyboardButton]] = []

    if question.options:
        for option in question.options:
            rows.append(
                [
                    InlineKeyboardButton(
                        text=_display_option_text(option, selected_options),
                        callback_data=f"q|{question.id}|{option.key}",
                    )
                ]
            )

    if question.multi_select:
        rows.append(
            [
                InlineKeyboardButton(
                    text=messages.MULTI_SELECT_DONE_BUTTON[1],
                    callback_data=f"q|{question.id}|done",
                )
            ]
        )

    return _build_inline_keyboard(rows)


def _display_option_text(option: Option, selected_options: Set[str]) -> str:
    if option.key in selected_options:
        return option.text
    return option.text


def consultation_keyboard() -> Optional[InlineKeyboardMarkup]:
    url = settings.consultation_url.strip()
    if not url:
        return None
    button = InlineKeyboardButton(text=messages.CONSULTATION_BUTTON_TEXT, url=url)
    return _build_inline_keyboard([[button]])

File: src/bot/keyboards.js
import { Markup } from 'telegraf';
import settings from '../config/settings.js';
import * as messages from './messages.js';

export function skillLevelKeyboard() {
  const rows = messages.SKILL_LEVEL_OPTIONS
    .filter(([, text]) => Boolean(text))
    .map(([callbackData, text]) => [Markup.button.callback(text, callbackData)]);
  return Markup.inlineKeyboard(rows);
}

export function startKeyboard() {
  return singleButtonKeyboard(messages.START_BUTTON);
}

export function singleButtonKeyboard([callbackData, text]) {
  return Markup.inlineKeyboard([[Markup.button.callback(text, callbackData)]]);
}

export function questionOptionsKeyboard(question, selectedOptions = new Set()) {
  const rows = [];

  if (Array.isArray(question?.options)) {
    for (const option of question.options) {
      const buttonText = option.text;
      rows.push([
        Markup.button.callback(buttonText, `q|${question.id}|${option.key}`)
      ]);
    }
  }

  if (question?.multiSelect) {
    rows.push([
      Markup.button.callback(messages.MULTI_SELECT_DONE_BUTTON[1], `q|${question.id}|done`)
    ]);
  }

  return Markup.inlineKeyboard(rows);
}

export function consultationKeyboard() {
  const url = (settings.consultationUrl || '').trim();
  if (!url) return null;
  return Markup.inlineKeyboard([[Markup.button.url(messages.CONSULTATION_BUTTON_TEXT, url)]]);
}
